dvc/repo/experiments/__init__.py

```diff
@@ -421,12 +434,42 @@ def _reproduce(self, executors: dict, jobs: Optional[int] = 1) -> dict:
 
         return result
 
-    def _collect_output(self, rev: str, executor: ExperimentExecutor):
-        logger.debug("copying tmp output from '%s'", executor.tmp_dir)
-        tree = self.scm.get_tree(rev)
-        for fname in tree.walk_files(tree.tree_root):
-            src = executor.path_info / relpath(fname, tree.tree_root)
-            copyfile(src, fname)
+    def _collect_output(self, executor: ExperimentExecutor):
+        """Copy (download) output from the executor tree into experiments
+        workspace.
+        """
+        from dvc.cache.local import _log_exceptions
+
+        logger.debug("Collecting output from '%s'", executor.tmp_dir)
+        dest_tree = self.exp_dvc.tree
+        src_tree = executor.tree
+
+        from_infos = []
+        to_infos = []
+        names = []
+        for from_info in executor.collect_output():
+            from_infos.append(from_info)
+            fname = from_info.relative_to(src_tree.path_info)
+            names.append(str(fname))
+            to_info = dest_tree.path_info / fname
+            to_infos.append(dest_tree.path_info / fname)
+            logger.debug(f"from '{from_info}' to '{to_info}'")
+        total = len(from_infos)
+
+        func = partial(
+            _log_exceptions(dest_tree.download, "download"),
+            dir_mode=dest_tree.dir_mode,
+            file_mode=dest_tree.file_mode,
+        )
+        with Tqdm(total=total, unit="file", desc="Downloading") as pbar:
+            func = pbar.wrap_fn(func)
+            # TODO: parallelize this, currently --jobs for repro applies to
+            # number of repro executors not download threads
+            with ThreadPoolExecutor(max_workers=1) as dl_executor:
```

Comment:
Will need to figure out better usage for this in the future, I think using `--jobs` for repro executors will probably end up changing, so at that point we can use it for download jobs the same way as w/other commands

Reply:
Maybe we should set some sensible default? Like max(2, cpu_count()/2)?